File: leetcode.com/python/785_Is_Graph_Bipartite.py
from collections import deque


class Solution(object):
    def isBipartite(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: bool
        """
        color = {}
        for node in range(len(graph)):
            if node not in color and graph[node]:
                queue = deque([node])
                color[node] = 0
                while queue:
                    currentNode = queue.popleft()
                    for neighbour in graph[currentNode]:
                        if neighbour not in color:
                            queue.append(neighbour)
                            color[neighbour] = color[currentNode] ^ 1
                        elif color[neighbour] == color[currentNode]:
                            return False
        return True

File: leetcode.com/swift/DS_ALGO_PRAC.playground/Sources/swift/785_Is_Graph_Bipartite.swift
import Foundation
// BFS
class Solution {
    func isBipartite(_ graph: [[Int]]) -> Bool {
        guard graph.count > 0 else {
            return false
        }
        for i in 0..<graph.count {
            var colorMap = [Int:String]()
            colorMap[i] = "RED"
            var nodeQueue = [(i, "RED")]
            while !nodeQueue.isEmpty {
                var (node, color) = nodeQueue.removeFirst()
                for neighbour in graph[node] {
                    if let neighbourColor = colorMap[neighbour] {
                        if neighbourColor == color {
                            return false
                        }
                    } else {
                        if color == "RED" {
                            nodeQueue.append((neighbour, "BLUE"))
                            colorMap[neighbour] = "BLUE"
                        } else {
                            nodeQueue.append((neighbour, "RED"))
                            colorMap[neighbour] = "RED"
                        }
                    }
                }
            }
        }
        return true
    }
}
